SparseGrids/tsgGridSequence.hpp

```diff
@@ -131,11 +133,44 @@ class GridSequence : public BaseCanonicalGrid{
         return cache;
     }
 
+    // The second argument should be the result of calling TasGrid::GridSequence::cacheBasisValues(x).
+    template<typename T>
+    std::vector<std::vector<T>> cacheBasisDerivatives(const T x[], std::vector<std::vector<T>> &value_cache) const {
```

Comment:
Many loops, checks, branching, special cases, division by zero problems, the need to pass in `value_cache` ...

Years ago, when I was studying computer science, my teachers were telling me that I should spend 15% of my time coding and 85% thinking about what I am about to code.

How about we replace this template by the following:
```
    template<typename T>
    std::vector<std::vector<T>> cacheBasisDerivatives(const T x[]) const {
        std::vector<std::vector<T>> cache(num_dimensions);
        for(int j=0; j<num_dimensions; j++){
            cache[j].resize(max_levels[j] + 1);
            T this_x = x[j];
            T b = 1.0;
            T s = 1.0;
            cache[j][0] = 0.0;
            if (max_levels[j] > 0) cache[j][1] = 1.0 / coeff[1];
            for(int i=2; i <= max_levels[j]; i++){
                s *= (this_x - nodes[i-1]);
                b *= (this_x - nodes[i-2]);
                s += b;
                cache[j][i] = s / coeff[i];
            }
        }
        return cache;
    }
```

Reply:
Looks good to me. I didn't immediately see that the product rule could be done recursively here.

Reply:
The Newton polynomials are inherently recursive, which will manifest in all aspects of the problem even when it is not immediately obvious. The recursion makes a big difference in speeding things up.